server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,378 @@
+                bucketOrd += 1;
+            }
+        }
+
+        mergeBuckets(mergeMap, initialNumBuckets);
+        if (deferringCollector != null) {
+            deferringCollector.mergeBuckets(mergeMap);
+        }
+    }
+
+    /**
+     * Returns the ordinal of the bucket whose centroid is closest to <code>val</code>, or -1 if there are no buckets.
+     */
+    private int getNearestBucket(double val) {
+        double smallest_distance = Double.POSITIVE_INFINITY;
+        int nearest_bucket = -1;
+        for(int i = 0; i < clusterCentroids.size(); i++){
+            double current_distance = Math.abs(clusterCentroids.get(i) - val);
+            if(current_distance < smallest_distance){
+                smallest_distance = current_distance;
+                nearest_bucket = i;
+            }
+        }
+        return nearest_bucket;
+    }
+
+
+    @Override
+    public InternalAggregation buildAggregation(long bucket) throws IOException {
+        assert bucket == 0;
+
+        if(cachedValues.size() < cacheLimit){
+            bucketCachedDocs();
+        }
+
+        consumeBucketsAndMaybeBreak((int) clusterCentroids.size());
+
+        long[] bucketOrdArray = new long[(int) clusterCentroids.size()];
+        for (int i = 0; i < clusterCentroids.size(); i++) {
+            bucketOrdArray[i] = i;
+        }
+
+        runDeferredCollections(bucketOrdArray);
+
+        List<InternalVariableWidthHistogram.Bucket> buckets = new ArrayList<>((int) clusterCentroids.size());
+        for (long i = 0; i < clusterCentroids.size(); i++) {
+            buckets.add(new InternalVariableWidthHistogram.Bucket(
+                clusterCentroids.get(i),
+                new InternalVariableWidthHistogram.Bucket.BucketBounds(clusterMins.get(i), clusterMaxes.get(i)),
+                bucketDocCount(i),
+                formatter,
+                bucketAggregations(i)));
+        }
+
+        // The contract of the histogram aggregation is that shards must return
+        // buckets ordered by key in ascending order
+        CollectionUtil.introSort(buckets, BucketOrder.key(true).comparator(this));
+
+        InternalVariableWidthHistogram.BucketInfo emptyBucketInfo = new InternalVariableWidthHistogram.BucketInfo(
+            buildEmptySubAggregations()
+        );
+
+
+        return new InternalVariableWidthHistogram(name, buckets, emptyBucketInfo, numBuckets, formatter,
+            pipelineAggregators(), metaData());
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+
+        InternalVariableWidthHistogram.BucketInfo emptyBucketInfo = new InternalVariableWidthHistogram.BucketInfo(
+            buildEmptySubAggregations()
+        );
+        return new InternalVariableWidthHistogram(name(), Collections.emptyList(), emptyBucketInfo, numBuckets, formatter,
+            pipelineAggregators(), metaData());
+    }
+
+    @Override
+    public void doClose() {
+        Releasables.close(cachedValues);
```

Comment:
better to do `Releasables.close(cachedValues, cluserMaxes, clusterMins, clusterCentroids, clusterSizes)`.